src/include/platform/NetworkCommissioning.h

```diff
@@ -0,0 +1,260 @@
+    uint8_t networkNameLen;
+    uint8_t ssidLen;
+    uint16_t channel;
+    uint8_t version;
+    uint64_t extendedAddress;
+    int8_t rssi;
+    uint8_t lqi;
+};
+
+using NetworkIterator            = Iterator<Network>;
+using WiFiScanResponseIterator   = Iterator<WiFiScanResponse>;
+using ThreadScanResponseIterator = Iterator<ThreadScanResponse>;
+
+// BaseDriver and WirelessDriver are the common interfaces for a network driver, platform drivers should not implement this
+// directly, instead, users are expected to implement WiFiDriver, ThreadDriver and EthernetDriver.
+namespace Internal {
+class BaseDriver
+{
+public:
+    /**
+     * @brief Returns maximum number of network configs can be added to the driver.
+     */
+    virtual uint8_t GetMaxNetworks() = 0;
+    /**
+     * @brief Returns an iterator for reading the networks, the user will always call NetworkIterator::Release.
+     */
+    virtual NetworkIterator * GetNetworks() = 0;
+
+    /**
+     * @brief Sets the status of the interface, this is an optional feature of a network driver.
+     */
+    virtual CHIP_ERROR SetEnabled(bool enabled) { return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE; }
+    /**
+     * @brief Returns the status of the interface, this is an optional feature of a network driver the driver will be enabled by
+     * default.
+     */
+    virtual bool GetEnabled() { return true; };
+
+    virtual ~BaseDriver() = default;
+};
+
+class WirelessDriver : public Internal::BaseDriver
+{
+public:
+    class ConnectCallback
+    {
+    public:
+        virtual void OnResult(Status commissioningError, CharSpan debugText, int32_t connectStatus) = 0;
+
+        virtual ~ConnectCallback() = default;
+    };
+
+    /**
+     * @brief Persist the network configuration. This function is expected to be called when CommissioningComplete event is fired.
+     */
+    virtual CHIP_ERROR CommitConfiguration() = 0;
+
+    /**
+     * @brief Revert the network configuration to the last committed one. This function is expected to be called when failsafe
+     * timeout reached.
+     */
+    virtual CHIP_ERROR RevertConfiguration() = 0;
+
+    virtual uint8_t GetScanNetworkTimeoutSeconds()    = 0;
+    virtual uint8_t GetConnectNetworkTimeoutSeconds() = 0;
+
+    virtual Status RemoveNetwork(ByteSpan networkId)                 = 0;
+    virtual Status ReorderNetwork(ByteSpan networkId, uint8_t index) = 0;
+
+    /**
+     * @brief Called to initiate a network join. callback->OnResult must be called, on both success and error. Callback can be
+     * called inside ConnectNetwork.
+     */
+    virtual void ConnectNetwork(ByteSpan networkId, ConnectCallback * callback) = 0;
+};
+} // namespace Internal
+
+class WiFiDriver : public Internal::WirelessDriver
+{
+public:
+    class ScanCallback
+    {
+    public:
+        /**
+         * Indicates the scan is finished, and accepts a iterator of networks discovered.
+         * - Iterator::Release must be called inside this function.
```

Comment:
It must be called *eventually*. It does not need to be called inside this function. It must *always* be called at some point, otherwise the next Scan call may run out of resources.

Reply:
I want to keep the assumption that the iterator will be fully consumed inside this callback.

Since we already calling this callback in thread-safe manner, the cluster can encode the response inside this callback. And there is no reason for the cluster to obtain the ownership of the iterator.

Having this assumption can make some platform use a bit simplier implementation (e.g. alloc some items on the stack if their event loop is intrgrated with the CHIP event loop.)